### PR TITLE
org.openjkdf2.OpenJKDF2

### PR DESCRIPTION
**Pull request by shinyquagsire23**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission
____

Notes for Flathub Approval:
- I'm the project maintainer, I've been doing regular Win64/macOS releases [here](https://github.com/shinyquagsire23/OpenJKDF2/releases) for a while and Flatpak looks like a reasonable distribution method for Linux.
- [openjkdf2.org](https://openjkdf2.org/) is also mine.
- OpenJKDF2 is a game engine reimplementation similar to OpenRCT2, so it requires original game assets. No copyrighted assets are included in the distributable.

Permissions rundown:
- It's an SDL2 app, so OpenGL/HID controllers seem to need `--device=all`.
- On first startup, the app will prompt the user to locate a mounted install disk, or an existing game installation from GOG/Steam.
- If a copy of the game is found, assets are copied to `$XDG_DATA_HOME/openjkdf2/`.
- Player profiles, settings and savegames are stored in `$XDG_DATA_HOME/openjkdf2/player/`.
- Normally, OpenJKDF2 can run in a portable mode if it detects assets in the working directory (ie, on a flash drive or shared with another OS). For now, this will require a manual permission override.
- Mods are frequently installed by users, so it is expected that `$XDG_DATA_HOME/openjkdf2/` will be modified externally by those users.

Notes for Flatpak Team:
- Really really appreciate that cross-compiling is seamless and works on ARM64.
- Kinda wish there was a better-documented way to not run the x86_64 compiler through QEMU and use a cross-compiler instead. But eh, I get why that might not be the most practical.

### CODE REVIEW

File: org.openjkdf2.OpenJKDF2.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+app-id: org.openjkdf2.OpenJKDF2
+runtime: org.freedesktop.Platform
+runtime-version: '21.08'
```
Comment (hfiguiere) [inline org.openjkdf2.OpenJKDF2.yml:3]:
```suggestion
runtime-version: '22.08'
```

File: org.openjkdf2.OpenJKDF2.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+sdk: org.freedesktop.Sdk
+
+command: openjkdf2-64
+
+modules:
+  - packaging/flatpak/glew/glew.json
+  - packaging/flatpak/glu/glu-9.json
```
Comment (hfiguiere) [inline org.openjkdf2.OpenJKDF2.yml:10]: please use the shared-modules for these two: https://github.com/flathub/shared-modules/

File: org.openjkdf2.OpenJKDF2.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  - name: openjkdf2-64
+    buildsystem: simple
+    build-commands:
+      - bash packaging/flatpak/build_linux64_flatpak.sh
+    sources:
+      - type: git
+        url: https://github.com/shinyquagsire23/OpenJKDF2
+        commit: 421a3d927e756ad47820893aea0bfacb6e749772
+
+cleanup-commands:
```
Comment (hfiguiere) [inline org.openjkdf2.OpenJKDF2.yml:22]: modules is usually the last section

File: org.openjkdf2.OpenJKDF2.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  - 'pip3 uninstall -y cogapp' # We only use this for building
+
+finish-args:
+  - --device=all # needed for gamepads
+  - --share=ipc  # needed for X11
+  - --share=network # needed for networking (Multiplayer/Valve GameNetworkingSockets)
+  - --socket=fallback-x11
+  - --socket=pulseaudio
+  - --filesystem=~/.local/share/openjkdf2:rw # Assets, mods, and saves are located here.
+  - --filesystem=host:ro # One-time asset installation prompts user for a disk (ie /run/media mounts) or an existing install (ie Steam, GOG).
```
Comment (hfiguiere) [inline org.openjkdf2.OpenJKDF2.yml:32]: this is too broad.

File: org.openjkdf2.OpenJKDF2.yml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+cleanup-commands:
+  - 'pip3 uninstall -y cogapp' # We only use this for building
+
+finish-args:
+  - --device=all # needed for gamepads
+  - --share=ipc  # needed for X11
+  - --share=network # needed for networking (Multiplayer/Valve GameNetworkingSockets)
+  - --socket=fallback-x11
+  - --socket=pulseaudio
+  - --filesystem=~/.local/share/openjkdf2:rw # Assets, mods, and saves are located here.
```
Comment (hfiguiere) [inline org.openjkdf2.OpenJKDF2.yml:31]: why don't you use XDG location instead?

File: org.openjkdf2.OpenJKDF2.yml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+app-id: org.openjkdf2.OpenJKDF2
```
Comment (hfiguiere) [inline org.openjkdf2.OpenJKDF2.yml:1]: openjkdf2.org is a placeholder. Is it own by the developer of that app?

File: org.openjkdf2.OpenJKDF2.yml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+modules:
+  - packaging/flatpak/glew/glew.json
+  - packaging/flatpak/glu/glu-9.json
+  - packaging/flatpak/python3-cogapp.json
+  
+  - name: openjkdf2-64
+    buildsystem: simple
+    build-commands:
+      - bash packaging/flatpak/build_linux64_flatpak.sh
```
Comment (hfiguiere) [inline org.openjkdf2.OpenJKDF2.yml:16]: by Linux 64, you mean any 64-bits architecture, right?

File: org.openjkdf2.OpenJKDF2.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    sources:
+      - type: git
+        url: https://github.com/shinyquagsire23/OpenJKDF2
+        commit: 421a3d927e756ad47820893aea0bfacb6e749772
+
+cleanup-commands:
+  - 'pip3 uninstall -y cogapp' # We only use this for building
```
Comment (hfiguiere) [inline org.openjkdf2.OpenJKDF2.yml:23]: install you should just `cleanup: '*'`. **in the module**

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by hfiguiere** [comment]
bot, build org.openjkdf2.OpenJKDF2

**Comment by hfiguiere** [comment]
```
builddir/files/share/appdata/org.openjkdf2.OpenJKDF2.appdata.xml: FAILED:
• tag-missing           : <release> required
Validation of files failed
```

File: org.openjkdf2.OpenJKDF2.yml (hunk 2 above)
Comment (shinyquagsire23) [reply org.openjkdf2.OpenJKDF2.yml:10]: Ah, I'd run into a certificate error with the glu-9 download, but it seems to be resolved now? Or it's cached somewhere and still broken I guess.

File: org.openjkdf2.OpenJKDF2.yml (hunk 7 above)
Comment (shinyquagsire23) [reply org.openjkdf2.OpenJKDF2.yml:16]: Yes, I test primarily on AArch64, but I also support x86_64. I guess 32-bit is technically supported though (I just don't test on it often) so I'll rename this.

File: org.openjkdf2.OpenJKDF2.yml (hunk 6 above)
Comment (shinyquagsire23) [reply org.openjkdf2.OpenJKDF2.yml:1]: I'm the only developer of the app and the domain is mine, yes.

File: org.openjkdf2.OpenJKDF2.yml (hunk 7 above)
Comment (hfiguiere) [reply org.openjkdf2.OpenJKDF2.yml:16]: we only support these two 64-bits arch.

I'm surprised that for once "linux64" doesn't equate x86_64. :-)

File: org.openjkdf2.OpenJKDF2.yml (hunk 2 above)
Comment (hfiguiere) [reply org.openjkdf2.OpenJKDF2.yml:10]: works fine everywhere else.

File: org.openjkdf2.OpenJKDF2.yml (hunk 4 above)
Comment (shinyquagsire23) [reply org.openjkdf2.OpenJKDF2.yml]: Appreciate the pushback on this, I moved to GtkFileChooserNative and the sandboxing had no issues.

**Comment by shinyquagsire23** [comment]
Updated based on feedback:

- GtkFileChooserNative and portals are used now, instead of `--filesystem=host:ro`.
- Use `--filesystem=xdg-data:rw` instead of `--filesystem=~/.local/share/openjkdf2:rw`.
- Cleanup moved to the cogapp module.
- Renamed `linux64` -> `linux` and `openjkdf2-64` -> `openjkdf2` to avoid confusion on i386 builds.
- SDK 22.08 instead of 21.08 (the getting started documentation still references 21.08 just FYI)

File: org.openjkdf2.OpenJKDF2.yml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+command: openjkdf2
+
+finish-args:
+  - --device=all # needed for gamepads
+  - --share=ipc  # needed for X11
+  - --share=network # needed for networking (Multiplayer/Valve GameNetworkingSockets)
+  - --socket=fallback-x11
+  - --socket=pulseaudio
+  - --filesystem=xdg-data:rw # Assets, mods, and saves are located here.
```
Comment (hfiguiere) [inline org.openjkdf2.OpenJKDF2.yml:14]: you don't need this. It's automatically provided (and persisted)

Comment (shinyquagsire23) [reply org.openjkdf2.OpenJKDF2.yml]: Fixed

**Comment by barthalion** [comment]
bot, build org.openjkdf2.OpenJKDF2

**Pull request closed without merging**